syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.grpc.examples.helloworld";
option java_outer_classname = "HelloWorldProto";
option objc_class_prefix = "HLW";

package helloworld;

// The greeting service definition.
service Greeter {
    // Sends a greeting
    rpc CreatePost (CreatePostRequest) returns (Post) {}

    rpc UpvoteDownvotePost (PostID) returns (Post) {}

    rpc RetrievePost (PostID) returns (Post) {}

    rpc CreateComment (CreateCommentRequest) returns (Comment) {}

    rpc UpvoteDownvoteComment (CommentID) returns (Comment) {}

    rpc CommentStream (ExpandCommentBranchRequest) returns (stream Comment) {}

    //rpc MonitorUpdate (stream ) 
}

// --------------- Data Model ---------------// 
// User
message User {
  string user_id = 1;
}

// Post
message PostID {
    string id = 1;
}

enum PostState {
    NORMAL = 0;
    LOCKED = 1;
    HIDDEN = 2;
}

message Post {
    PostID id = 0;
    string title = 1;
    string text = 2;
    oneof url {
        string video = 3;
        string img = 4;
    }
    optional string author = 5;
    int32 score = 6;
    PostState state = 7;
    string pub_date = 8;
    Subreddit subreddit = 9;
}

message CreatePostRequest {
    string title = 1;
    string text = 2;
    oneof url {
        string video = 3;
        string img = 4;
    }
    optional string author = 5;
    string pub_date = 8;
    Subreddit subreddit = 9;
}

// Comment
message CommentID {
    string id = 1;
}

enum CommentState {
    NORMAL = 0;
    HIDDEN = 1;
}

message Comment {
    CommentID id = 9;
    string author = 0;
    int32 score = 1;
    CommentState state = 2;
    string pub_date = 3;
    oneof parent {
        Post parent_post_id = 4;
        CommentID parent_comment_id = 5;
    }
    bool hasReplies = 6;
}

message CreateCommentRequest {
    string author = 0;
    string pub_date = 1;
    CommentState state = 2;
    oneof parent {
        Post parent_post_id = 3;
        CommentID parent_comment_id = 4;
    }
}

message MostUpvotedCommentRequest {
    PostID post = 0;
    int32 number = 1;
}

message ExpandCommentBranchRequest {
    CommentID comment = 0;
    int32 number = 1;
}

// Subreddit
enum SubredditState {
    PUBLIC = 0;
    PRIVATE = 1;
    HIDDEN = 2;
}

message Subreddit {
    string name = 0;
    SubredditState state = 1;
    string pub_date = 2;
    repeated string tags = 3;
}